### DISCUSSION TITLE
`<ScrollRestoration>` with scrolling container other than `window`

### DISCUSSION THREAD

**Discussion by david-crespo** [remix-run/react-router#9495] [Proposals]
I briefly [discussed](https://discord.com/channels/770287896669978684/1033046089034641488/1033046089034641488) this with @brophdawg11 on Discord and he suggested starting a discussion here. He raised some useful concerns:

> I'm sure the component could be altered to support a more advanced nested scroll use-case, but it introduces edge cases that would need to be thought through.  Like what if when you come back to that location the scrollElement is no longer present for some reason?  falling back on window isn't going to be right since the scroll position is relative to the no-longer-rendered element etc.

## The problem

If you use CSS grid to define a layout where only one pane scrolls rather than the entire page, `<ScrollRestoration>` does not work because it only uses `window.scrollY` and `window.scrollTo(_, _)`.

https://github.com/remix-run/react-router/blob/5f3cfb7ac2ad823ecdf69069332567e35ad795c6/packages/react-router-dom/index.tsx#L1018-L1110

## [CodeSandbox demo](https://codesandbox.io/s/react-router-scroll-restoration-non-window-lyj044?file=/src/App.tsx)

![2022-10-24-scroll-restore-demo](https://user-images.githubusercontent.com/3612203/197569387-abc366ac-5865-447a-9663-92e76f8b7d38.gif)

## Very Bad patch fix

In my app at work, I'm working around this by using [patch-package](https://github.com/ds300/patch-package) to simply use `document.querySelector('#content-pane')` instead of `window`. This is a little scary, but it seems to work quite well. `#content-pane` is in layouts so we know it will more or less always be there.

```diff
diff --git a/node_modules/react-router-dom/dist/index.js b/node_modules/react-router-dom/dist/index.js
index aeec512..b617027 100644
--- a/node_modules/react-router-dom/dist/index.js
+++ b/node_modules/react-router-dom/dist/index.js
@@ -796,6 +796,19 @@ function useScrollRestoration(_temp3) {
   let matches = useMatches();
   let navigation = useNavigation(); // Trigger manual scroll restoration while we're active
 
+  // HACK: Scroll restoration doesn't work out of the box (issue #1155) because
+  // the container that scrolls is the content pane, one of the page-level grid
+  // cells. So instead of scrolling window, we pull that element directly by ID
+  // and use it instead. In the unlikely event it's not there, fall back to
+  // window (and likely do nothing).
+  const getScrollTarget = () => document.querySelector('#content-pane') || window;
+
+  // window has scrollY but normal elements have scrollTop
+  const getScrollY = () => {
+    const el = getScrollTarget();
+    return el ? el.scrollY || el.scrollTop : undefined;
+  }
+
   React.useEffect(() => {
     window.history.scrollRestoration = "manual";
     return () => {
@@ -806,7 +819,7 @@ function useScrollRestoration(_temp3) {
   useBeforeUnload(React.useCallback(() => {
     if (navigation.state === "idle") {
       let key = (getKey ? getKey(location, matches) : null) || location.key;
-      savedScrollPositions[key] = window.scrollY;
+      savedScrollPositions[key] = getScrollY();
     }
 
     sessionStorage.setItem(storageKey || SCROLL_RESTORATION_STORAGE_KEY, JSON.stringify(savedScrollPositions));
@@ -825,7 +838,7 @@ function useScrollRestoration(_temp3) {
   }, [storageKey]); // Enable scroll restoration in the router
 
   React.useLayoutEffect(() => {
-    let disableScrollRestoration = router == null ? void 0 : router.enableScrollRestoration(savedScrollPositions, () => window.scrollY, getKey);
+    let disableScrollRestoration = router == null ? void 0 : router.enableScrollRestoration(savedScrollPositions, () => getScrollY(), getKey);
     return () => disableScrollRestoration && disableScrollRestoration();
   }, [router, getKey]); // Restore scrolling when state.restoreScrollPosition changes
 
@@ -837,7 +850,7 @@ function useScrollRestoration(_temp3) {
 
 
     if (typeof restoreScrollPosition === "number") {
-      window.scrollTo(0, restoreScrollPosition);
+      getScrollTarget().scrollTo(0, restoreScrollPosition);
       return;
     } // try to scroll to the hash
 
@@ -856,8 +869,7 @@ function useScrollRestoration(_temp3) {
       return;
     } // otherwise go to the top on new locations
 
-
-    window.scrollTo(0, 0);
+    getScrollTarget().scrollTo(0, 0);
   }, [location, restoreScrollPosition, preventScrollReset]);
 }
 
```

## More general approach

Obviously hard coding an element doesn't make sense for the real library. Here are some ways I can picture it working:

```ts
<ScrollRestoration getScrollTarget={() => document.querySelector('#content-pane')} />
```

and in the library:

```
const getScrollTarget = () => props.getScrollTarget() || window;
```

Calling `document.querySelector` in React is usually not a thing we want to do, so maybe it should be a ref instead. I think this would require putting `ScrollRestoration` next to your scroll container so you can get the ref:

```ts
const ContentPane = () => {
  const ref = useRef(null)
  return (
    <>
      <ScrollRestoration scrollTargetRef={ref} /> 
      <div ref={ref}>...</div>
    </>
  )
}
```

In both cases I think falling back to `window` should be ok, because generally it will do nothing. Though maybe that's only true in my use case because we've made sure the whole page doesn't scroll. So maybe it's safer not to fall back, and instead just don't do anything if the ref is null or if the selector function comes back null.

**Comment by Antonio-Bennett** [comment]
Dude you are a saviour!! Plus helped me to learn about patch-package which I did not know before :) worked like a charm!

**Comment by ryanflorence** [comment]
I've built a hook to restore scroll of individual elements:

```jsx
function SomeComp() {
  useElementScrollRestoration("content-pane");

  return (
    <div id="content-pane">
      {/* ... */}
    </div>
  )
}
```

And then the inside of that looks a whole lot like our code inside of `<ScrollRestoration>`. They ought to be able to share almost all of the same code.

I'd love to add this hook. If you or someone else wants to take this on I'd love to offer a few tips on how I think I'd implement it.

**Reply by david-crespo** [reply]
Sure, I'd happily make the PR. How do you propose to get around the `scrollY` vs. `scrollTop` thing? Do you think it should also accept a ref instead of an ID? That seemed the more Officially Correct way to me.

**Reply by david-crespo** [reply]
https://github.com/remix-run/react-router/pull/9573

**Reply by litera** [reply]
Would be a better hook if it was called `useScrollRestorationRef` and then set the ref on the element:
```
function SomeComp() {
  const ref = useScrollRestorationRef(null);

  return (
    <div ref={ref}>
      ...
    </div>
  )
}
```

**Comment by esmalleydev** [comment]
Ran into this issue as well, would love to be able to pass `<ScrollRestoration>` a ref.

**Comment by NestorRV** [comment]
I have a similar situation to the one showed in the [CodeSandbox demo](https://codesandbox.io/s/react-router-scroll-restoration-non-window-lyj044?file=/src/App.tsx). I'm using a big `<div>` as a grid to build a grid of _cards_. If I scroll down (_going through_ the div) and click back and forward, the position is not retained...

**Comment by david-crespo** [comment]
My above hack still works for scrolling to top on forward navigation, but when I upgraded from 6.4.2 to 6.6.1, I noticed that scroll restore on back nav doesn't work anymore — it always scrolls to top. What's especially odd is I put in a `console.log` at the spot where it is supposed to be restoring, and it seems to be doing what I expect it to do, i.e., the scroll restoration number is present and it's a reasonable number, and the target element is also there with a reasonable `offsetHeight`. In any case, happy to pick #9573 back up or redo it when @ryanflorence has guidance.

**Comment by shish** [comment]
I am also hitting a practically identical issue (my app uses flexbox for static header / scrollable body / static footer) -- being able to pass a ref to ScrollRestoration sounds ideal :)

**Comment by shish** [comment]
> I'm sure the component could be altered to support a more advanced nested scroll use-case, but it introduces edge cases that would need to be thought through. Like what if when you come back to that location the scrollElement is no longer present for some reason?

Using a ref seems like it would fix this

- If ref is undefined, the developer is intending that we use the window position
- If ref is defined, the developer is intending that we use the element position
  - if the ref is populated, then we save and restore that element
  - if the ref is empty, we do nothing

Does that work? Are there other edge cases I'm not thinking of?

("What if we want to support multiple scrolled elements side-by-side?" seems like a potential case, but my answer right now is "this change is designed to support the common use-case of having a single scrolling element, and only that case; supporting multiple elements would probably require more significant API and internal changes than adding a single optional parameter")

**Comment by jrnail23** [comment]
@ryanflorence, would you mind sharing your `useElementScrollRestoration` implementation?

**Reply by david-crespo** [reply]
The implementation in here works well: https://github.com/remix-run/react-router/pull/10468

I'm using it as a patch with [`patch-package`](https://github.com/ds300/patch-package), here is the patch (filename `react-router-dom+6.11.2.patch`):

```diff
diff --git a/node_modules/react-router-dom/dist/index.d.ts b/node_modules/react-router-dom/dist/index.d.ts
index 8ff0e6e..3d8bd3e 100644
--- a/node_modules/react-router-dom/dist/index.d.ts
+++ b/node_modules/react-router-dom/dist/index.d.ts
@@ -138,6 +138,7 @@ export declare const Form: React.ForwardRefExoticComponent<FormProps & React.Ref
 export interface ScrollRestorationProps {
     getKey?: GetScrollRestorationKeyFunction;
     storageKey?: string;
+    elementRef?: React.RefObject<HTMLElement>;
 }
 /**
  * This component will emulate the browser's scroll restoration on location
@@ -218,6 +219,7 @@ export declare function useFetchers(): Fetcher[];
 declare function useScrollRestoration({ getKey, storageKey, }?: {
     getKey?: GetScrollRestorationKeyFunction;
     storageKey?: string;
+    elementRef?: React.RefObject<HTMLElement>;
 }): void;
 export { useScrollRestoration as UNSAFE_useScrollRestoration };
 /**
diff --git a/node_modules/react-router-dom/dist/index.js b/node_modules/react-router-dom/dist/index.js
index c2a7ae7..499adb5 100644
--- a/node_modules/react-router-dom/dist/index.js
+++ b/node_modules/react-router-dom/dist/index.js
@@ -597,11 +597,13 @@ if (process.env.NODE_ENV !== "production") {
 function ScrollRestoration(_ref7) {
   let {
     getKey,
-    storageKey
+    storageKey,
+    elementRef,
   } = _ref7;
   useScrollRestoration({
     getKey,
-    storageKey
+    storageKey,
+    elementRef,
   });
   return null;
 }
@@ -884,6 +886,22 @@ function useFetchers() {
 }
 const SCROLL_RESTORATION_STORAGE_KEY = "react-router-scroll-positions";
 let savedScrollPositions = {};
+
+const getScrollTarget = (elementRef) =>
+  elementRef ? elementRef.current : window;
+
+const getScrollY = (elementRef) => {
+  const el = getScrollTarget(elementRef);
+  // window has scrollY but normal elements have scrollTop
+  return !el ? 0 : el.scrollY || el.scrollTop;
+}
+
+const scrollY = (elementRef, y) => {
+  const el = getScrollTarget(elementRef);
+  if (el) {
+    el.scrollTo(0, y);
+  }
+}
 /**
  * When rendered inside a RouterProvider, will restore scroll positions on navigations
  */
@@ -891,7 +909,8 @@ let savedScrollPositions = {};
 function useScrollRestoration(_temp3) {
   let {
     getKey,
-    storageKey
+    storageKey,
+    elementRef,
   } = _temp3 === void 0 ? {} : _temp3;
   let {
     router
@@ -914,12 +933,12 @@ function useScrollRestoration(_temp3) {
   usePageHide(React.useCallback(() => {
     if (navigation.state === "idle") {
       let key = (getKey ? getKey(location, matches) : null) || location.key;
-      savedScrollPositions[key] = window.scrollY;
+      savedScrollPositions[key] = getScrollY(elementRef);
     }
 
     sessionStorage.setItem(storageKey || SCROLL_RESTORATION_STORAGE_KEY, JSON.stringify(savedScrollPositions));
     window.history.scrollRestoration = "auto";
-  }, [storageKey, getKey, navigation.state, location, matches])); // Read in any saved scroll locations
+  }, [storageKey, getKey, elementRef, navigation.state, location, matches])); // Read in any saved scroll locations
 
   if (typeof document !== "undefined") {
     // eslint-disable-next-line react-hooks/rules-of-hooks
@@ -936,9 +955,9 @@ function useScrollRestoration(_temp3) {
     // eslint-disable-next-line react-hooks/rules-of-hooks
 
     React.useLayoutEffect(() => {
-      let disableScrollRestoration = router == null ? void 0 : router.enableScrollRestoration(savedScrollPositions, () => window.scrollY, getKey);
+      let disableScrollRestoration = router == null ? void 0 : router.enableScrollRestoration(savedScrollPositions, () => getScrollY(elementRef), getKey);
       return () => disableScrollRestoration && disableScrollRestoration();
-    }, [router, getKey]); // Restore scrolling when state.restoreScrollPosition changes
+    }, [router, getKey, elementRef]); // Restore scrolling when state.restoreScrollPosition changes
     // eslint-disable-next-line react-hooks/rules-of-hooks
 
     React.useLayoutEffect(() => {
@@ -949,7 +968,7 @@ function useScrollRestoration(_temp3) {
 
 
       if (typeof restoreScrollPosition === "number") {
-        window.scrollTo(0, restoreScrollPosition);
+        scrollY(elementRef, restoreScrollPosition);
         return;
       } // try to scroll to the hash
 
@@ -969,8 +988,8 @@ function useScrollRestoration(_temp3) {
       } // otherwise go to the top on new locations
 
 
-      window.scrollTo(0, 0);
-    }, [location, restoreScrollPosition, preventScrollReset]);
+      scrollY(elementRef, 0);
+    }, [location, restoreScrollPosition, preventScrollReset, elementRef]);
   }
 }
 /**
```

**Reply by jrnail23** [reply]
Thanks, @david-crespo, I stumbled upon that PR shortly after digging into yours.
I'm pretty much copy-pasting it into my app right now to try it out (I'm a bit leery of going the `patch-package` route, as I'm not very familiar with it yet).  Hopefully that PR gets merged soon so I can get rid of the copy-pasta!.

**Reply by david-crespo** [reply]
patch-package can be a little annoying when the hook changes and you have to update the patch. I think you’re wise to go the copy paste route. TBH I might switch to that.

**Reply by jrnail23** [reply]
@david-crespo, I've been pulling my hair out all day trying to get this working, but now I'm running into a similar problem that you had -- when it's time to restore the scroll, the ref to my scroll container is null.  Did you find a fix for that issue?

**Reply by jrnail23** [reply]
FWIW, I got it working using the id of the scroll container instead of using a ref.

**Reply by jrnail23** [reply]
It may be a good idea to support both refs and element IDs in the final implementation, because of this issue.  It can be such a pain in the ass to deal with refs like that, particularly when their elements are deeper in the tree, and as such, often not rendered yet.

**Reply by david-crespo** [reply]
I feel like if you can get working with an ID but not a ref, something must be a little off. Your scrolling container should be high in the tree and not conditionally rendered. In the issue I mentioned in the PR, my problem was not that the ref was undefined, but that the div wasn't at its full height yet.

**Reply by jrnail23** [reply]
I think my issue is probably related to our scrolling container itself -- it's a component provided by our design system library that supports sticky headers/footers, etc., and I don't really have any control over its innards.

**Comment by jrnail23** [comment]
Cross-posting from https://github.com/remix-run/react-router/pull/10468#issuecomment-1559694079

> Just some feedback from trying this on my own: maybe consider supporting elementId as well as elementRef (presumably mutually exclusive) -- I was able to get this PR's general approach working in my app using elementId, but still having trouble getting it to work with elementRef. In my case, elementRef.current is still empty when it's time to restore the scroll position, so it never gets restored.
> 
> I'm really having a hard time understanding how to ensure the scroll container's ref gets populated by the time the scroll restore happens. If ScrollRestoration is placed higher in the React tree than your scroll container, is that even possible? What am I missing here?

I also wanted to add a specific use case we have, to hopefully get feedback and maybe inform future design decisions:

We've got a scrollable list of items, and clicking one navigates to a detail page.  That detail page has a "Back to List" link, where when clicked, should navigate back to the list view, restoring the scroll position to the initially clicked item.  While we also want scroll restore to work on the browser's back button and have its state survive a reload, that "Back to List" link isn't just a `navigate(-1)` because that detail page also has stuff you can do (e.g., flip through images using a carousel, or select sub-details to view) that would push new locations (via query string variables), but still stay on the same page (pathname).

For the record, that initial scrollable list is a component itself, with things like optional sticky headers/footers, so `window.scrollTo` isn't appropriate, and while it's not at the top of the React tree, it's reasonably close.  As I mentioned above, I just can't seem to bend the implementation in #10468 to my will to get this working at all, nevermind the "Back to List" link, which likely isn't directly possible with that PR's approach.  

FWIW, we made all this work using React Router v5, although I'm not particularly proud of that implementation, TBH.  The key for us with RRv5 was that the `Link` component's `to` prop accepted a function, whereby when one of the list item links are clicked, we'd capture the container's scroll position in a `previousLocation` prop in location state on navigate.  I was able get this mostly working by hacking together a customized version of `useLinkClickHandler`, but it's pretty ugly, and I'd love to find a cleaner way to do this.

**Comment by martpie** [comment]
Adding my use-case to this thread. One of the most requested feature for my Electron app (museeks.io) is scroll restoration between views. Some views may have thousand of elements in them, and losing context when switching from one to the next is tedious.

Why not use body/window? Because we have a fixed header and footer, as most desktop applications, and it doesn't make sense to have an overflow on the whole page.

It would awesome to see an official solution here :D

<img width="1012" alt="Screenshot 2023-05-30 at 14 13 33" src="https://github.com/remix-run/react-router/assets/1311607/5f381cae-c450-4145-8e54-2dcc48c404e9">

**Comment by jnorvell** [comment]
I am also having this issue. My problem, which has also been state above, is that `<ScrollRestoration>` only seems to work at the window level, and if the scrollable content is further down the hierarchy, it does not seem to work properly.  I'm using v6.15.0.  For sure that are many fringe cases that may not be practically supportable by a general library, but basic cases like scrollable components within a page seems like a reasonable request.

**Reply by jnorvell** [reply]
The current version is now v6.20.0.   Just checking to see if there is an official solution for this issue scheduled for a specific release.

**Comment by brophdawg11** [comment]
FWIW @kentcdodds just released this package yesterday https://github.com/epicweb-dev/restore-scroll

**Reply by Virthuss** [reply]
Sadly I'm having issues with  https://github.com/epicweb-dev/restore-scroll
Is there any plan to implement a clean way to easily restore container's scrolling position for 6.21.3 or following ? An official solution would truly be awesome, either trough a hook or with the basic ScrollRestoration component :)

**Reply by ivan-kleshnin** [reply]
Does not work in RR-7 "framework mode".

**Comment by bartley20** [comment]
Thanks for the workaround! It's really unfortunate that ScrollRestoration doesn't support specifying an element rather than the window

**Comment by CestDiego** [comment]
+1 to needing resotre scroll position of scrollable divs that aren't window.

**Reply by CestDiego** [reply]
Moreover it should be able to restore both: To a place in window where the div is showing and also the place inside the div.  :)

**Comment by cristianrgreco** [comment]
I also ran into this. I ended up migrating from react-router to tanstack router which supports this. Only took a few minutes and it works as you'd expect: https://tanstack.com/router/v1/docs/framework/react/migrate-from-react-router

**Comment by lensbart** [comment]
I’ve been able to make it work similarly to the `patch-package` approach. I’ll include the patch below for those who want to give it a try. Differences from the approach above:

* Introduces a prop for `getScrollContainer` (dev & prod, incl. types)
* Changes a couple of additional lines of code that I think were missed

[edit: see [gist](https://gist.github.com/lensbart/ffe8cc4ead36d1561b15b77ccd5473d0) below]

**Comment by lensbart** [comment]
To add to the above: using this patch, you can write the following:

```tsx
// root.tsx
export default (_: Route.ComponentProps) => (
  <html>
    <head>{/* ... */}</head>
    <body>
      {/* ... */}
      <ScrollRestoration
        getScrollContainer={() => document.querySelector("main.overflow-auto")}
      />
      <Scripts />
    </body>
  </html>
)
```

**Comment by lensbart** [comment]
For those who are interested in trying this out with the latest version of react-router (7.6.0), I’ve updated the patch: https://gist.github.com/lensbart/ffe8cc4ead36d1561b15b77ccd5473d0

**Reply by ivan-kleshnin** [reply]
@lensbart thank you for the work 🙏  Have you considered to create a PR? Or release it as a standalone library? I don't get why this baseline functionality has to be added by hacks.

**Reply by lensbart** [reply]
Yes, I have considered it. But prior to opening a PR I checked with the React Router team [on the Remix Discord](https://discord.com/channels/770287896669978684/940264701785423992/1369368887879270571), and @brophdawg11 [replied](https://discord.com/channels/770287896669978684/1369381782029598804/1374385598739251201) that a PR was unlikely to get merged in the short term, due to the team being busy with other priorities.

In the meantime, [this gist](https://gist.github.com/lensbart/d2f0c3a049f602fc4a826a2b462dcdbc) might help you apply the patch above to an upgraded react-router dependency. I’ve used it to successfully migrate from 7.6.0 to 7.6.1.

I agree that this is a suboptimal situation and I would much prefer to upstream my improvements into react-router itself.

**Comment by lensbart** [comment]
I updated the patch to `react-router@7.6.2` (see [this gist](https://gist.github.com/lensbart/7310ad3066195a1f755289c1740d9bf3)) using the [script](https://gist.github.com/lensbart/d2f0c3a049f602fc4a826a2b462dcdbc) mentioned above.

In light of [the recently-announced governance model](https://github.com/remix-run/react-router/blob/main/GOVERNANCE.md), this proposal can be considered stage 1 when 

> two or more SC members indicate interest/support for the idea as a valuable addition to React Router

@ryanflorence has [expressed](https://github.com/remix-run/react-router/discussions/9495#discussioncomment-4082505) his interest in the proposal, albeit for a different approach (`useElementScrollRestoration(...)` instead of `<ScrollRestoration getScrollContainer={...} />`).

@brophdawg11 has [suggested](https://discord.com/channels/770287896669978684/940264701785423992/1369381436754497577) to provide a `patch-package`, although he has warned that it’s too early to open a PR, and implied that other proposals have a higher priority.

It would be great to hear if there is appetite in the Remix team to move this proposal to stage 1. I’m open to rewriting the implementation a bit (e.g. to make it work as a hook), if that would be welcomed. Then I can open a PR and we can discuss the specifics of the approach. If the existing patch is more or less in the right direction, much of the work has already been done and it won’t take long to let people try out this approach.

Thanks in advance!

**Comment by Crauzer** [comment]
I really hope this gets fixed soon, I also have a website with a scrolling container and It's really annoying having to deal with this issue.

**Comment by lensbart** [comment]
@brophdawg11 could you check with the RR team if there is any way to move this proposal forward? There is [a simple proposal that works well](https://github.com/remix-run/react-router/discussions/9495#discussioncomment-13384150), but reapplying a patch with every RR update is quite tedious. Many thanks

**Comment by HyeryongChoi** [comment]
**+1,** with a use case that hasn't been mentioned yet: native **WebViews** where the scroll root isn't `window`.

We ship a React Router app inside native iOS/Android WebViews. In several in-app browsers (Instagram iOS, some Android WebViews) the scroll root is `document.body` / `document.documentElement` / `document.scrollingElement`, not `window`. Since `ScrollRestoration` hardcodes `window.scrollY` / `window.scrollTo`, restoration is fully broken there:

- **POP (back/forward):** the saved value is `0` (the scroll lived on `body`, not `window`), so nothing restores.
- **PUSH:** the previous page's `body` offset is retained by the WebView and `window.scrollTo(0, 0)` doesn't clear it, so the new page starts mid-scroll.

A configurable scroll target like [@lensbart's `getScrollContainer` patch](https://gist.github.com/lensbart/7310ad3066195a1f755289c1740d9bf3) looks like the right shape for this — in principle you'd point it at the real scroll root:

```tsx
<ScrollRestoration getScrollContainer={() => document.scrollingElement} />
```

One caveat from our own experience: which element actually holds the scroll varies across these WebViews (`body` vs `documentElement` vs `scrollingElement`), so in practice we had to *detect* the active scroll root at runtime rather than assume `scrollingElement`. I haven't verified lensbart's patch against these specific WebViews, so I can't claim it works out of the box — but a single configurable target that can return the detected root would be enough to cover this case.

Flagging it as another class of problem worth SC interest: unlike the grid/flexbox cases above, this isn't an app design choice — it's the runtime environment. The same code that works in a normal browser silently breaks inside a WebView, with no supported fix short of reimplementing `ScrollRestoration` (which is what we currently do).
